tests/pmap_test.py

```diff
@@ -747,6 +749,25 @@ def testPswapaxes(self):
     expected = onp.swapaxes(x, 0, 2)
     self.assertAllClose(ans, expected, check_dtypes=False)
 
+  def testReshardInput(self):
+    if xla_bridge.device_count() < 6:
+      raise SkipTest("testReshardInput requires 6 devices")
+    # Manually construct a ShardedDeviceArray with the wrong sharding for the
+    # subsequent pmap
+    shard_shape = (3,2)
+    shard = np.arange(np.prod(shard_shape)).reshape(shard_shape)
+    bufs = [xla.device_put(shard, d) for d in xla_bridge.devices()[:4]]
+    aval = ShapedArray((6,4), shard.dtype)
+    indices = [(slice(0,3), slice(0,2)),
+               (slice(0,3), slice(2,4)),
+               (slice(3,6), slice(0,2)),
+               (slice(3,6), slice(2,4))]
+    arr = pxla.ShardedDeviceArray(aval, indices, bufs)
+
+    r = pmap(lambda x: x + 1)(arr)
+    self.assertAllClose(r, arr + 1, check_dtypes=True)
+    self.assertEqual(len(r.device_buffers), 6)
```

Comment:
I think we can also assert something about the indices of `r`. Perhaps instead of asserting the length of `r.device_buffers` assert the value of `d.logical_indices`. Then one can read from this test the sharding behavior.